### PR TITLE
JGRP-2295 use InetAddress.getAllByName() for DNS_PING A record lookups

### PR DESCRIPTION
**Pull request by rcernich**
https://issues.jboss.org/browse/JGRP-2295
https://issues.jboss.org/browse/JGRP-2292

Signed-off-by: rcernich <rcernich@redhat.com>

### CODE REVIEW

File: src/org/jgroups/protocols/dns/DNS_PING.java [hunk 1]
```diff
@@ -36,6 +36,9 @@
     @Property(description = "DNS query for fetching members")
     protected String dns_query;
 
+    @Property(description = "Append default suffix of .svc.cluster.local to specified query")
+    protected boolean append_default_suffix = true;
```
[the remark this answers is not in the document]
Comment (rcernich) [reply src/org/jgroups/protocols/dns/DNS_PING.java]: Is it okay to leave it blank by default, or will that disrupt users too much?

File: src/org/jgroups/protocols/dns/OSDNSResolver.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+import javax.naming.NamingException;
+import javax.naming.directory.DirContext;
+import java.lang.invoke.MethodHandles;
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import java.util.ArrayList;
+import java.util.List;
+
+class OSDNSResolver extends DefaultDNSResolver {
```
[the remark this answers is not in the document]
Comment (rcernich) [reply src/org/jgroups/protocols/dns/OSDNSResolver.java]: I can do that, and nuke the SRV record bit (why I chose to specialize the DefaultDNSResolver).

**Comment by rcernich** [comment]
I'm still not sure how the user would specify which resolver they want to use.

File: src/org/jgroups/protocols/dns/OSDNSResolver.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (slaskawi) [reply src/org/jgroups/protocols/dns/OSDNSResolver.java]: +1, just throw something there...

File: src/org/jgroups/protocols/dns/DNS_PING.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (slaskawi) [reply src/org/jgroups/protocols/dns/DNS_PING.java]: Yes, exactly. IMO this should be left empty by default, meaning that no suffix is being added.

**Comment by slaskawi** [comment]
> I'm still not sure how the user would specify which resolver they want to use.

As I suggested in the ticket, I think it would be best to have a parameter with class name. That parameter would be annotated with `@Property`, so that it can be changed in the configuration. Here's a pseudo-code example:

```
@Property(description = "A DNS Resolver implementation. In you want OS resolver pick blah...")
protected String dns_resolver_class = "org.jgroups.protocols.dns.DefaultDNSResolver";
...
    @Override
    public void init() throws Exception {
        ....
        dns_resolver = Class.forName(dns_resolver_class).newInstance();
    }
```

WDYT?

**Comment by slaskawi** [comment]
If the OS resolver gets adopted widely, we can make it the default one.

**Comment by rcernich** [comment]
I recommended the A/SRV split because the OS resolver needs to use DnsContext for SRV, which doesn't make much sense.  So simplifying, one resolver for SRV, one for A.  Having said that, you can still get away with A for service records by prefixing your query with _endpoints, so... (or maybe that's what I do for the OS resolver so it works for both?).

**Comment by rcernich** [comment]
I redid this after reviewing the specs for k8s and openshift dns resolvers.  I don't think it makes much sense to use DnsContextDir for A records at all, but it must be used for SRV records.  Given that, I simply modified the existing DefaultDNSResolver so that it used InetAddress.getAllByName() for A records and DnsContextDir for SRV records.

I also removed the query suffix completely, as it allows users to specify whatever they need for the query, without having it modified.  This will mean that existing users will need to update their query strings by adding .svc.cluster.local to their query strings.  That said, if they're using A records, with a headless service for their clustering, they should be able to simply use the cluster name, as the native lookup should resolve correctly (as openshift/k8s should be adding the proper search suffixes to the pod's /etc/resolve.conf file).  This should simplify things greatly for users.

I could not figure out how to modify the unit tests and it looks like they're disabled in the pom.  That said, I was thinking about implementing a mock NameService to resolve the A records appropriately.

**Comment by slaskawi** [comment]
> I redid this after reviewing the specs for k8s and openshift dns resolvers. I don't think it makes much sense to use DnsContextDir for A records at all, but it must be used for SRV records. Given that, I simply modified the existing DefaultDNSResolver so that it used InetAddress.getAllByName() for A records and DnsContextDir for SRV records.

That works absolutely fine for k8s use case. However, the initial scope of this protocol was broader than k8s and I would prefer to keep it that way. Just to give you an example, there are some Amazon folks in the community that use `DNS_PING` with their own DNS Server. I know very little about their setup but if they specified the `dns_address` and queried the `A` records, this PR would potentially break their application, since the new `A` records resolver completely ignores `dns_address` parameter. Moreover, those folks will have no way to bring the old behavior back. 

So my proposed solution remains the same - let's introduce another `DNSResolver` implementation. Feel free to either extend the `DefaultDNSResolver` or just implement `DNSResolver` and nuke `SRV` records. Once this is done, please make your new resolver the default one, and introduce a mechanism to change the resolver. This way both worlds should be happy. All folks based on k8s will get a new, better DNS resolver and the Amazon folks will be able to switch back to the old implementation. This way we will make the migration also easy for them. 

> I also removed the query suffix completely, as it allows users to specify whatever they need for the query, without having it modified. This will mean that existing users will need to update their query strings by adding .svc.cluster.local to their query strings. That said, if they're using A records, with a headless service for their clustering, they should be able to simply use the cluster name, as the native lookup should resolve correctly (as openshift/k8s should be adding the proper search suffixes to the pod's /etc/resolve.conf file). This should simplify things greatly for users.

I totally agree. However, just for visibility, //cc @ryanemerson @osnipso @luck3y @iankko @pferraro @rhusar 

> I could not figure out how to modify the unit tests and it looks like they're disabled in the pom. That said, I was thinking about implementing a mock NameService to resolve the A records appropriately.

Oh, yes, testing the `InetAddress` is very tricky. All I can suggest is to extract `InetAddress.getAllByName()` call into a separate method (e.g. `OSDNSResolver#getIPAddresses()`) and override it during the test. That's a nasty trick but it might work.

**Comment by slaskawi** [comment]
I should probably find a moment early tomorrow and I'll try to create a
sample commit to show you what I mean. I think it will be simpler that way
;)

>

**Comment by slaskawi** [comment]
> I should probably find a moment early tomorrow and I'll try to create a
sample commit to show you what I mean. I think it will be simpler that way
;)

Let me reiterate... The only requirement I have here is to be able to bring the old resolver back (if I want).

**Comment by rcernich** [comment]
@slaskawi, I think I have something figured out...s/b transparent.

**Comment by rcernich** [comment]
Okay, legacy behavior is preserved if the user specifies a dns_address, otherwise the new impl is used.  This should be transparent to the user, with the exception that the default suffix has been removed (i.e. folks relying on that will need to add it to their dns_query).

**Comment by rcernich** [comment]
@slaskawi, I'm done, out of cycles.  Feel free to merge or dump it off onto somebody else.

**Comment by rcernich** [comment]
@slaskawi, hold off a minute.  I did not revert the updates to the properties' text.

**Comment by slaskawi** [comment]
@rcernich Ok, let me pull it locally tomorrow morning and check it out. Great work. Thanks for all the effort Rob!

**Comment by rcernich** [comment]
@slaskawi, okay, s/b gtg.

**Comment by belaban** [comment]
Can JGRP-2292 and JGRP-2295 be resolved now?

**Comment by slaskawi** [comment]
@belaban Yes! Please close both tickets. I merged this manually because I also fixed the tests and updated the docs. All in `master` now.

@rcernich Thanks a lot for all the heavy weightlifting!

**Pull request merged**